### PR TITLE
Let drvRegistry to override remote plugin

### PR DESCRIPTION
**Pull request by mavenugo**
Addresses https://github.com/docker/docker/issues/29473

drvRegistry isnt aware if a plugin is v1 or v2. Plugin-v2 provides a way
for user to disable and remove plugins. But unfortunately, there isnt
any api to advertise the removal to drvRegistry. Hence there is no way
to handle "docker plugin rm" of installed plugin. In order to support
the case of "docker plugin install x" followed by "docker plugin rm x"
followed by reinstalling of plugin "docker plugin install x",
drvRegistry must allow overriding any existing plugin with the same
name. The protection in plugin infra will prevent willful override of
existing plugin.

### CODE REVIEW

**Comment by mavenugo** [comment]
ping @anusha-ragunathan @aboch @yongtang

**Comment by sanimej** [comment]
@mavenugo If we can't support `docker plugin rm` in a meaningful way may we should disable `docker plugin rm` for network plugins instead of adding code to support fake  `docker plugin rm` ` and subsequent `docker plugin install` ?

File: controller.go [hunk 1]
```diff
@@ -476,12 +479,23 @@ func (c *controller) ID() string {
...
-	for _, i := range getInitializers(c.cfg.Daemon.Experimental) {
-		if i.ntype == "remote" {
-			continue
+	c.drvRegistry.WalkDrivers(func(name string, driver driverapi.Driver, capability driverapi.Capability) bool {
+		if driver.IsBuiltIn() {
+			drivers = append(drivers, name)
 		}
-		drivers = append(drivers, i.ntype)
-	}
+		return false
```
Comment (anusha-ragunathan) [inline controller.go]: Not related to this change: Looks like all `WalkDrivers` return `false` so that `DrvRegistry`'s `WalkDrivers` can iterate through all drivers. Given that why not simply return, rather than return a bool?

File: controller.go [hunk 2]
```diff
@@ -476,12 +479,23 @@ func (c *controller) ID() string {
...
+	})
+	return drivers
+}
+
+func (c *controller) BuiltinIPAMDrivers() []string {
```
Comment (anusha-ragunathan) [inline controller.go]: How and where is this new method invoked?

Comment (mavenugo) [inline controller.go]: We added the `BuiltInDriver` API for docker/docker to make use of. But we missed adding it for IPAM drivers. This API is added so that docker/docker can make use of it. Infact, this is required to fix one of the plugin related issues that I will push a PR in docker/docker once this PR is merged and vendored.

File: controller.go (hunk 1 above)
Comment (mavenugo) [inline controller.go]: the function is defined in such a way that if a callback function returns a `true`, the walk will be terminated. In our case, we dont want to break the walk and hence we always return false.

File: drvregistry/drvregistry.go [hunk 3]
```diff
@@ -164,10 +164,10 @@ func (r *DrvRegistry) RegisterDriver(ntype string, driver driverapi.Driver, capa
 	}
 
 	r.Lock()
-	_, ok := r.drivers[ntype]
+	dd, ok := r.drivers[ntype]
 	r.Unlock()
 
-	if ok {
+	if ok && dd.driver.IsBuiltIn() {
```
Comment (anusha-ragunathan) [inline drvregistry/drvregistry.go]: We should atleast add a comment about pluginv2 behaviour, where the older instance of the plugin doesnt get removed from drvRegistry and re-installing a new instance of the plugin overwrites the driverData here.

Comment (anusha-ragunathan) [inline drvregistry/drvregistry.go]: For pluginv1, this will cause confusion since there is no removal of the older instance. So at a time, registering 2 instances of the same plugin will not cause an error anymore.

File: controller.go (hunk 1 above)
Comment (anusha-ragunathan) [inline controller.go]: At the moment, there's no callback that returns `true`.

Comment (anusha-ragunathan) [inline controller.go]: Also, its strange for an iterator function to return a bool.

File: controller.go (hunk 2 above)
Comment (anusha-ragunathan) [inline controller.go]: ok

File: driverapi/driverapi.go [hunk 4]
```diff
@@ -74,6 +74,9 @@ type Driver interface {
 
 	// Type returns the the type of this driver, the network type this driver manages
 	Type() string
+
+	// IsBuiltIn returns true if it is a built-in driver
+	IsBuiltIn() bool
```
Comment (aboch) [inline driverapi/driverapi.go]: Have you considered adding a `builtin` flag into the `initializer` structure instead of adding this new method ?

```
type initializer struct {
	fn      drvregistry.InitFunc
	ntype   string
+	builtin bool
}
```

Adding the new API is not really needed for this fix.

Comment (aboch) [inline driverapi/driverapi.go]: Hmmm, it looks like you need this info outside of `BuiltInDrivers()`. Then your current change is probably what is needed.

Comment (mavenugo) [inline driverapi/driverapi.go]: Yes. I wanted to use it from DrvRegistry.

File: controller.go (hunk 1 above)
Comment (mavenugo) [inline controller.go]: How can we instruct the iterator to break if the callback prefers to stop the iteration ?
BTW, this is not related to this change and we can continue to discuss this offline :)

File: drvregistry/drvregistry.go (hunk 3 above)
Comment (mavenugo) [inline drvregistry/drvregistry.go]: @anusha-ragunathan  Yes. If we dont do it, it wont work at all for Plugin-V2. Since V2 will deprecate V1, it is easier to document this user configuration error than to support the supported plugin-v2. WDYT ?

Comment (anusha-ragunathan) [inline drvregistry/drvregistry.go]: Yes, adding a fat comment describing the behaviour is fine by me.

**Comment by aboch** [comment]
LGTM
